### PR TITLE
Ticket2124: Delete multiple device screens

### PR DESCRIPTION
**Pull request by DominicOram**
### Description of work

Allowed the user to select and delete multiple screens in the edit window.

### To test

https://github.com/ISISComputingGroup/IBEX/issues/2124

To test:
* Confirm that you can select multiple screens in the edit view.
* Confirm that the edit box is disabled when multiple screens are selected
* Confirm that you can delete when multiple screens are selected
* Confirm that pointing to a read only instrument you can only create local device screens

Note that there was quite a bit of refactoring here, general testing of the device screens view would be good.

---

#### Code Review

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/GUI-Coding-Conventions)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit tests in place? Are the unit tests small and test the a class in isolation?
- [ ] Are there [system tests](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/System-Testing-with-RCPTT) in place? Do they test a minimal set of functionality and leave the gui as close as possible to its original state?
- [ ] Did any existing system test break as a result of the current changes? 
- [ ] Have the changes been documented in the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev). If so, do they describe the changes appropriately?
- [ ] Has the [manual system tests spreadsheet](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Manual-system-tests) been updated?
- [ ] If an OPI has been modified, does it conform to the [style guidelines](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/OPI-Creation)? There is a script called `check_opi_format.py` to help with this.

### Functional Tests

- [ ] Do changes function as described? Add comments below that describe the tests performed.
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has developer documentation been updated if required?

### Final Steps
- [ ] Reviewer has moved the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev) entry for this ticket in the "Changes merged into master" section

### CODE REVIEW

File: base/uk.ac.stfc.isis.ibex.ui.devicescreens.tests/src/uk/ac/stfc/isis/ibex/ui/devicescreens/tests/DeviceScreensDescriptionViewModelTest.java [hunk 1]
```diff
@@ -134,55 +135,97 @@ public void constructor_produces_deep_copy_of_description() {
...
     @Test
     public void clearing_the_selected_screen_unsets_name_key_etc() {
         // Act
-        viewModel.setSelectedScreen(-1);
+        viewModel.setSelectedScreens(null);
 
         // Assert
-        assertEquals("", viewModel.getCurrentName());
+        assertEquals("", viewModel.getName());
         assertEquals("", viewModel.getCurrentKey());
```
Comment (AdrianPotter) [inline base/uk.ac.stfc.isis.ibex.ui.devicescreens.tests/src/uk/ac/stfc/isis/ibex/ui/devicescreens/tests/DeviceScreensDescriptionViewModelTest.java]: Why has "Current" been removed from other methods but not this one? On examination of EditDeviceScreensDescriptionViewModel, it doesn't track previous keys. Would using standard naming of getKey() and setKey() be more appropriate?

Comment (AdrianPotter) [inline base/uk.ac.stfc.isis.ibex.ui.devicescreens.tests/src/uk/ac/stfc/isis/ibex/ui/devicescreens/tests/DeviceScreensDescriptionViewModelTest.java]: Why has "Current" been removed from other methods but not this one? On examination of EditDeviceScreensDescriptionViewModel, it doesn't track previous keys. Would using standard naming of getKey() and setKey() be more appropriate?

File: base/uk.ac.stfc.isis.ibex.ui.devicescreens.tests/src/uk/ac/stfc/isis/ibex/ui/devicescreens/tests/DeviceScreensDescriptionViewModelTest.java [hunk 2]
```diff
@@ -134,55 +135,97 @@ public void constructor_produces_deep_copy_of_description() {
...
-        assertEquals("", viewModel.getCurrentDescription());
-        assertEquals(null, viewModel.getSelectedScreen());
+        assertEquals("", viewModel.getDescription());
+    }
+
+    @Test
+    public void setting_multiple_screens_unsets_name_key_etc() {
```
Comment (AdrianPotter) [inline base/uk.ac.stfc.isis.ibex.ui.devicescreens.tests/src/uk/ac/stfc/isis/ibex/ui/devicescreens/tests/DeviceScreensDescriptionViewModelTest.java]: I think the team standard is now to use GIVEN__WHEN__THEN naming. From the point of view of behaviour driven development, this test suggests that the name, key and description MUST be set to blank when multiple screens are selected. Is this true or has blank string been selected as a sensible option? Although you might leave the test content the same, make sure that the name reflects the desired behaviour, not the specific implementation.

This comment repeats as necessary for all other new tests.

File: base/uk.ac.stfc.isis.ibex.ui.devicescreens.tests/src/uk/ac/stfc/isis/ibex/ui/devicescreens/tests/DeviceScreensDescriptionViewModelTest.java [hunk 3]
```diff
@@ -134,55 +135,97 @@ public void constructor_produces_deep_copy_of_description() {
...
+        // Act
+        List<DeviceDescriptionWrapper> first = viewModel.getScreens().subList(0, 1);
+        viewModel.setSelectedScreens(first);
+
+        // Assert
+        assertEquals(first.get(0), viewModel.getTargetScreen());
     }
 
     @Test
     public void changing_the_name_propagates() {
```
Comment (AdrianPotter) [inline base/uk.ac.stfc.isis.ibex.ui.devicescreens.tests/src/uk/ac/stfc/isis/ibex/ui/devicescreens/tests/DeviceScreensDescriptionViewModelTest.java]: More poor test naming. I know you haven't touched these but could I appeal to your philanthropic nature to improve all the tests in the class?

File: base/uk.ac.stfc.isis.ibex.ui.devicescreens/src/uk/ac/stfc/isis/ibex/ui/devicescreens/models/EditDeviceScreensDescriptionViewModel.java [hunk 4]
```diff
@@ -101,73 +115,43 @@ public EditDeviceScreensDescriptionViewModel(DeviceScreensModel deviceScreensMod
...
 
         canWriteRemote = deviceScreensModel.isCanWriteRemote();
-        setPersistenceEnabled();
 
         // From the description create a list of devices
-        devices = new ArrayList<>();
+        ArrayList<DeviceDescriptionWrapper> screens = new ArrayList<>();
         for (DeviceDescription d : deviceScreensModel.getDeviceScreensDescription().getDevices()) {
-            devices.add(new DeviceDescriptionWrapper(d, this.provider));
+            screens.add(new DeviceDescriptionWrapper(d, this.provider));
```
Comment (AdrianPotter) [inline base/uk.ac.stfc.isis.ibex.ui.devicescreens/src/uk/ac/stfc/isis/ibex/ui/devicescreens/models/EditDeviceScreensDescriptionViewModel.java]: Are we on Java 8 now? Can we use a stream here?

File: base/uk.ac.stfc.isis.ibex.ui.devicescreens/src/uk/ac/stfc/isis/ibex/ui/devicescreens/models/EditDeviceScreensDescriptionViewModel.java [hunk 5]
```diff
@@ -383,16 +373,9 @@ public String getCurrentKey() {
      *            the new value to set
      */
     public void setCurrentKey(String key) {
```
Comment (AdrianPotter) [inline base/uk.ac.stfc.isis.ibex.ui.devicescreens/src/uk/ac/stfc/isis/ibex/ui/devicescreens/models/EditDeviceScreensDescriptionViewModel.java]: Other methods have lost their "Current". This probably should too unless there's a good reason not to

File: base/uk.ac.stfc.isis.ibex.ui.devicescreens/src/uk/ac/stfc/isis/ibex/ui/devicescreens/models/EditDeviceScreensDescriptionViewModel.java [hunk 6]
```diff
@@ -412,60 +394,36 @@ private void updateCurrentKey(String newKey) {
...
+     * Add a blank screen to the list of screens.
      */
     public void addScreen() {
-        List<DeviceDescriptionWrapper> oldList = new ArrayList<>(devices);
+        List<DeviceDescriptionWrapper> newList = new ArrayList<>(devices);
 
         DefaultName namer = new DefaultName("Screen", "_", false);
         List<String> names = new ArrayList<>();
-        for (DeviceDescriptionWrapper d : oldList) {
+        for (DeviceDescriptionWrapper d : newList) {
```
Comment (AdrianPotter) [inline base/uk.ac.stfc.isis.ibex.ui.devicescreens/src/uk/ac/stfc/isis/ibex/ui/devicescreens/models/EditDeviceScreensDescriptionViewModel.java]: Another opportunity to use a stream here?

File: base/uk.ac.stfc.isis.ibex.ui.devicescreens/src/uk/ac/stfc/isis/ibex/ui/devicescreens/models/EditDeviceScreensDescriptionViewModel.java [hunk 7]
```diff
@@ -562,16 +520,22 @@ public boolean canWriteRemote() {
...
+     * Deletes the screens and selects the previous item.
      */
-    public boolean getPersistenceEnabled() {
-        return previousPersistenceEnabled;
-    }
+    public void deleteSelectedScreens() {
+        int newIndex = devices.indexOf(selectedScreens.get(0)) - 1;
+        if (newIndex < 0) {
+            newIndex = 0;
+        }
```
Comment (AdrianPotter) [inline base/uk.ac.stfc.isis.ibex.ui.devicescreens/src/uk/ac/stfc/isis/ibex/ui/devicescreens/models/EditDeviceScreensDescriptionViewModel.java]: Surely just use Math.max(n,0)?

File: base/uk.ac.stfc.isis.ibex.ui.devicescreens/src/uk/ac/stfc/isis/ibex/ui/devicescreens/models/EditDeviceScreensDescriptionViewModel.java (hunk 4 above)
Comment (DominicOram) [inline base/uk.ac.stfc.isis.ibex.ui.devicescreens/src/uk/ac/stfc/isis/ibex/ui/devicescreens/models/EditDeviceScreensDescriptionViewModel.java]: No, we can't switch :( Can't remember the exact reason off the top of my head, Matt looked into it.

Comment (DominicOram) [inline base/uk.ac.stfc.isis.ibex.ui.devicescreens/src/uk/ac/stfc/isis/ibex/ui/devicescreens/models/EditDeviceScreensDescriptionViewModel.java]: I've asked him to put a wiki page up about the reasoning etc

**Pull request merged**